ipa-core/src/cli/playbook/streaming.rs

```diff
@@ -0,0 +1,283 @@
+use std::{
+    io::BufRead,
+    pin::Pin,
+    task::{Context, Poll, Waker},
+};
+
+use bytes::Bytes;
+use futures::Stream;
+
+use crate::{
+    error::BoxError,
+    helpers::BytesStream,
+    sync::{Arc, Mutex},
+};
+
+/// Trait for submitting inputs as streams, rather than reading everything
+/// in memory. Should provide better performance for very large inputs.
+trait StreamingSubmission {
+    /// Spits itself into `count` instances of [`BytesStream`].
+    fn into_byte_streams(self, count: usize) -> Vec<impl BytesStream>;
+}
+
+/// Round-Robin strategy to read off the provided buffer
+/// and distribute them. Inputs is expected to be hex-encoded
+/// and delimited by newlines. The output streams will have
+/// run-length encoding, meaning that each element will have
+/// a 2 byte length prefix added to it.
+struct RoundRobinSubmission<R>(R);
+
+impl<R: BufRead> RoundRobinSubmission<R> {
+    pub fn new(read_from: R) -> Self {
+        Self(read_from)
+    }
+}
+
+/// One individual stream that reads from the shared input for [`RoundRobinSubmission`]
+struct RoundRobinStream<R> {
+    /// Points to the shared state.
+    state: Arc<Mutex<State<R>>>,
+    /// Unique identifier of this stream
+    idx: usize,
+}
+
+impl<R: BufRead> Stream for RoundRobinStream<R> {
+    type Item = Result<Bytes, BoxError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut state = self.state.lock().unwrap();
+        // it can be made more efficient by using atomics to check index.
+        // but probably it does not matter for this code as it is not on the hot path
+        if state.idx == self.idx {
+            let poll = state.read_next();
+            if poll.is_ready() {
```

Comment:
does read_next only return Poll::Ready?

Reply:
no, but only if it is ready, we want to advance the index